src/Data/Deque/Mutable/Linear.hs

```diff
@@ -0,0 +1,182 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE LinearTypes #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+
+-- | Mutable Linear Deque
+--
+-- This module provides a pure interface to a mutable deque.
+--
+-- It is designed to be imported qualfied:
+--
+-- > import qualfied Data.Deque.Mutable.Linear as Deque
+module Data.Deque.Mutable.Linear
+  (
+  -- * Allocation
+    alloc
+  , fromList
+  -- * Querying
+  , size
+  , length
+  , isFull
+  , peekRight
+  , peekLeft
+  -- * Modification
+  , addLeft
+  , addRight
+  , popLeft
+  , popRight
+  , map
+  -- * Consumption
+  , toList
+  )
+where
+
+import qualified Data.Array.Mutable.Linear as Array
+import Data.Unrestricted.Linear
+import Prelude.Linear hiding (length, map)
+import qualified Prelude
+import GHC.Stack
+
+
+-- # Types
+-------------------------------------------------------------------------------
+
+data Deque a where
+  Deque :: Int -> Maybe (Int,Int) -> !(Array.Array a) %1-> Deque a
+  -- This is: Deque size (Maybe (left ptr,right ptr)) array
```

Comment:
There is a slightly simpler representation, where you only store the number of elements in the queue and the starting index. It always wraps around to the beginning at the end of the array, so `(start + count) % size`  would give the position of the next element.

So only `Int -> Int -> Array a` should suffice, I think.

Reply:
Sigh, yes, I'll use the simpler representation.